### PR TITLE
[SPARK-33814][SS] Provide preferred locations for stateful operations without reported state store locations

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This proposes to provide preferred locations for stateful operations even there is no reported state store locations.

### Why are the changes needed?

Stateful operators in SS provides preferred locations on the previous batches if any. However, if there is no previous batch to follow, Spark possibly schedules stateful tasks in inefficient distribution. In local test, Spark has scheduled stateful tasks to a few ones among all executors.

This is because when the first batch takes payload from latest offsets, this batch possibly finishes very quick. An executor might be assigned more than one task because the executor finishes previous task very quickly and becomes available again.

Generally this is not a problem. It doesn't mater these tasks are evenly distributed or not, because the tasks are finished very quickly. But for SS stateful tasks, next batches will choose task locations of previous batch. So this is why this is an issue only for SS.

As stateful operations probably need to maintain large state stores, it means large memory usage. A skew state store distribution possibly causes one or few executors to have memory usage times more than others. So it means the streaming query could be OOM on these executors, even there are other executors are still available for state stores.

It is better we distribute stateful tasks across all executors evenly. This patch proposes to provide preferred locations for stateful operations even there is no reported state store locations.

Summary:

1. This proposes to leverage existing preferred locations to have evenly initial state store locations. We already leverage it to have state store on same locations in existing implementation.
2. It is required to avoid uneven resource (memory, disk) requirement on executors.
3. It is also required to have stable state store locations across executors with locality setting.
4. Looks no other feasible approaches without making major change. Any major change can also possibly be regression for non-streaming jobs.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit test. Manual test on testing cluster.

### CODE REVIEW

**Comment by viirya** [comment]
cc @HeartSaVioR @HyukjinKwon @dongjoon-hyun

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:217]: If I understand correctly, we can simply do the round robin and achieve the same thing. 

Initialize the new immutable array or list from `context.getExecutorIds()` (if `context.getExecutorIds()` is also immutable this is also not necessary as well), and have a variable for index while would move by one per assignment (like `++idx % count`).

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:217]: Btw, what we'd like to achieve is trying to distribute state across all executors, then this PR looks to miss accounting executors containing active state. We'll want to also increase the count for the case `stateStoreLocs` is available as well. Probably better than simple round robin.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreRDD.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreRDD.scala:46]: Same here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:217]: Good point. I will update this.

**Comment by HyukjinKwon** [comment]
cc @zsxwing and @xuanyuanking too FYI

**Comment by viirya** [comment]
retest this please

**Review by HeartSaVioR — APPROVED** [review]
+1

**Comment by HeartSaVioR** [comment]
Will merge once new build passes. Retriggering...

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by viirya** [comment]
retest this please

**Comment by viirya** [comment]
Thanks @HeartSaVioR

**Comment by viirya** [comment]
retest this please

**Comment by zsxwing** [comment]
IMO, this looks a hacky approach to workaround a Spark task scheduler issue. Have we tried to improve Spark's task scheduler which will be a general improvement for all cases that miss preferred locations? For example, reading files from a cloud storage such as S3 has the same issue. Right?

**Comment by viirya** [comment]
> IMO, this looks a hacky approach to workaround a Spark task scheduler issue. Have we tried to improve Spark task scheduler which will be a general improvement for all cases that miss preferred locations? For example, reading files from a cloud storage such as S3 has the same issue. Right?

This is an issue for stateful operations because the expected high cost for maintaining multiple states in same executor. Because we prefer same executor of previous streaming batch as state location, so once the first batch chooses an bad state distribution, it could cause performance issue in later batches.

More, it becomes more severe for Structured Streaming because the first batch usually takes very quick time to finish, when it takes payload from latest offsets.

I think it is not an issue for Spark task scheduler for general tasks.

**Comment by viirya** [comment]
retest this please

**Comment by zsxwing** [comment]
Can you explain it a little bit why Spark cannot distribute the tasks evenly in the cluster? It would help me understand why this is not a problem for general tasks.

**Comment by viirya** [comment]
> Can you explain it a little bit why Spark cannot distribute the tasks evenly in the cluster? It would help me understand why this is not a problem for general tasks.

I ran some streaming queries with stateful operation recently. When the first batch takes payload from latest offsets, this batch possibly finishes very quick. An executor might be assigned more than one task because the executor finishes previous task very quickly and becomes available again.

Generally this is not a problem. It doesn't mater these tasks are evenly distributed or not, because the tasks are finished very quickly. But for SS stateful tasks, next batches will choose task locations of previous batch. So this is why this is an issue only for SS.

**Comment by viirya** [comment]
BTW, for stateful task's parent task is shuffle, if there is an empty partition which doesn't run a task in a batch, the next batch cannot find state location as preferred location so it will choose arbitrarily any executor. It also could make the state to be run in an executor there is already other state running.

**Comment by viirya** [comment]
retest this please

**Comment by zsxwing** [comment]
> When the first batch takes payload from latest offsets, this batch possibly finishes very quick. An executor might be assigned more than one task because the executor finishes previous task very quickly and becomes available again.

I see. This makes sense. But why do we need to avoid this?

> This is an issue for stateful operations because the expected high cost for maintaining multiple states in same executor.

What's the cost did you mean? The execution memory used by states?

It would be great if you can explain your case and what issue you would like to solve in the PR description.

Ideally, we should let the Spark task scheduler to do its work rather than doing the task scheduling work in SS because we don't have the full context of the executors. For example, this PR has to assume each executor has the same capability, while the task scheduler knows more about slow and fast executors.

**Comment by viirya** [comment]
> I see. This makes sense. But why do we need to avoid this?
> What's the cost did you mean? The execution memory used by states?
> It would be great if you can explain your case and what issue you would like to solve in the PR description.

To avoid skew memory usage on an executor. Yes, it is mainly for memory. For streaming queries that store large states, memory usage is severe. I will update the PR description to make it more clear.

> Ideally, we should let the Spark task scheduler to do its work rather than doing the task scheduling work in SS because we don't have the full context of the executors. For example, this PR has to assume each executor has the same capability, while the task scheduler knows more about slow and fast executors.

Preferred location doesn't replace the task scheduler, it is just a suggestion and task scheduler can choose to use it or not. For example we already asked later batch to schedule tasks on same executors that store states in previous batch. This is how the preferred locations work, isn't?

This PR doesn't assume executor capacity but suggests the task scheduler to evenly distribute statuful tasks across executors if possible, when no store location is available.

**Comment by viirya** [comment]
ping @HeartSaVioR @zsxwing Any more comments on this?

**Comment by HeartSaVioR** [comment]
I have to agree that this approach isn't ideal one, as the logic is blind on executors' status and just try to distribute stateful tasks evenly on executors. (And it's not guaranteed task scheduler follows the requests.) Probably the logic needs to update to reflect the actual state distribution per every batch, as the logic wouldn't know how task scheduler finally makes the decision.

But I also have to agree that there doesn't look to be other feasible approaches without making major change.

> Ideally, we should let the Spark task scheduler to do its work rather than doing the task scheduling work in SS because we don't have the full context of the executors. For example, this PR has to assume each executor has the same capability, while the task scheduler knows more about slow and fast executors.

Same applies to the task scheduler. Task scheduler doesn't have the full context of the characteristics on SS (preferred locations are not an enforcement), and given the cost of reloading state 
 (retrieving the file"s" from remote file system, and extracting the compression, and loading to the memory) is not trivial compared to the ideal micro-batch execution time, locality is no longer just a guidance.

Probably the point here is the view of the cost - whether it's ignorable or not, compared to the actual execution. Kafka data source has the same characteristic (Kafka client and unread fetched data is cached in executor) but probably less costly. Assuming the large state, it's going to be no longer ignorable.

If we want to draw the ideal picture here, IMO my ideal picture is to pin executors and force these executors to serve these stateful tasks on the lifetime of the query. It's ideal to guarantee these stateful tasks never have to reload the state unless crash. This would not be ideal if the application runs multiple queries where batch and streaming are mixed and streaming queries have longer trigger interval hence the chance to be idle. Either the query should wait to be assigned to the executor, or executor should be allowed to be idle for the query.

**Comment by zsxwing** [comment]
I'm wondering whether this would solve the issue. There are so many factors involved. I feel the initial even task distribution (assuming the executors are free so the task scheduler will respect the preferred locations) would become uneven quickly after some micro batches, caused by, such as, uneven partitions, executor lost, concurrent queries, etc... Did you verify that this would make a real workload that didn't work before become working?

If the issue is about memory, a low memory state store implementation such as https://github.com/qubole/spark-state-store is a better solution.

> If we want to draw the ideal picture here, IMO my ideal picture is to pin executors and force these executors to serve these stateful tasks on the lifetime of the query. It's ideal to guarantee these stateful tasks never have to reload the state unless crash. This would not be ideal if the application runs multiple queries where batch and streaming are mixed and streaming queries have longer trigger interval hence the chance to be idle. Either the query should wait to be assigned to the executor, or executor should be allowed to be idle for the query.

I think the answer to your ideal picture is continuous processing. Although we have not invested resources on continuous processing for a while, it's unlikely we would pin executors in micro batch since micro batch trades latency for throughput.

**Comment by viirya** [comment]
> I'm wondering whether this would solve the issue. There are so many factors involved. I feel the initial even task distribution (assuming the executors are free so the task scheduler will respect the preferred locations) would become uneven quickly after some micro batches, caused by, such as, uneven partitions, executor lost, concurrent queries, etc... Did you verify that this would make a real workload that didn't work before become working?

During running stateful streaming queries recently, it caused some troubles by bad initial locations of stores. I agree that this is not ideal, but it solves the problem I saw during I tested SS recently. It is simple and shouldn't have bad impact/regression to SS queries. Combined with task locality configuration, it makes SS queries more stable in my local test.

> If the issue is about memory, a low memory state store implementation such as https://github.com/qubole/spark-state-store is a better solution.

The problem is, we have a built-in in memory store. And, I don't think Rocksdb-based state store is the answer to all cases. Even with Rocksdb-based store, is it good to have skew stores on few executors? Then local disk space might be the next issue.

**Comment by viirya** [comment]
I put the summary of my points in the PR description. IMHO, I do not see a strong reason to object this change. This does not have back-compatibility issue and also does not change user-facing behavior. This does not change any API.

Actually without non-trivial task locality setting, this is a no-op because of Spark task scheduler's characteristics. Again, preferred location is only a hint to Spark task scheduler.

As one point mentioned (https://github.com/apache/spark/pull/30812#issuecomment-750544363) already is, there is no other feasible approach to deal with the issue of unstable state store locations across batches and skew state store locations without major change, and the issue is harmful to stateful streaming queries.

Please consider this patch so we may move things forward. If there is other feasible approach we can take, please share and I will be happy to take. Thanks.

**Comment by viirya** [comment]
Hi @zsxwing, would you mind to let me know that you are still against this change or you are okay to have this? Thanks a lot.

**Comment by viirya** [comment]
IMHO, per previous discussion, there are minor concerns and questions, and I think I and @HeartSaVioR replied. I don't see there is strong objection. If any, please let me know.

So if no further comments, and this is open for a while and already got approval from SS committer, I will assume this is okay to merge. I will wait for few days and merge this before the end of this week.

**Review by HeartSaVioR — CHANGES_REQUESTED** [review]
(There's no good way to take back my +1 on Github UI except this. Sigh. Sorry about this.)

Could we please defer the discussion after the release of Spark 3.1.0? We don't have to rush for this, as it's unlikely included into Spark 3.1.0. (3.2.0 is likely several months later.) Most people (including me) were on PTO and now have to focus on verification of the release.

I believe the discussion has been valuable - there's a consensus that the approach is clearly a workaround and also some mood that we'd like to have an ideal fix, which could take time. 

In addition, we don't seem to think about the various cases, including restoring query. For example, round-robin is probably a good start for the streaming query as we can't predict the future, but once the query ran for a while and we know there's a skew, round-robin might not be a good distribution (even in terms of memory usage on executors for state). That would happen when we restart the streaming query from checkpoint. That is an existing problem as well since Spark task scheduler has no idea of the state, but at least task scheduler would leverage another factors.

My apologize, but let's take one step back and revisit whether this approach is suitable for all (at least what we can imagine) cases. Let me take back my +1 for now.

**Comment by viirya** [comment]
ok for me. Thanks @HeartSaVioR

**Pull request closed without merging**
